[Core] Fix null reference accessing target framework monikers

mrward opened the pull request:
Creating a .NET Core console project, then adding a new non-SDK style
C# library project, it was not possible to open the Edit References
dialog for the library project. The IDE log would have a null
reference exception:

```
System.NullReferenceException: Object reference not set to an instance of an object
  at System.Collections.Immutable.ImmutableArray`1[T].ThrowNullRefIfNotInitialized () [0x00000] in <6518d0e60ca84f6b915c07341a031a32>:0
  at System.Collections.Immutable.ImmutableArray`1[T].GetEnumerator () [0x00007] in <6518d0e60ca84f6b915c07341a031a32>:0
  at MonoDevelop.PackageManagement.PackageManagementCanReferenceProjectExtension.CanReferenceProject
```

The problem was that the project's TargetFramework setter is not
called when adding a new project to an existing solution so the
project's TargetFrameworkMonikers array was not initialized. This is
now handled.

Fixes VSTS #973344 - Unable to open Edit References dialog for new
project

mrward committed:
[Core] Fix null reference accessing target framework monikers

Creating a .NET Core console project, then adding a new non-SDK style
C# library project, it was not possible to open the Edit References
dialog for the library project. The IDE log would have a null
reference exception:

System.NullReferenceException: Object reference not set to an instance of an object
  at System.Collections.Immutable.ImmutableArray`1[T].ThrowNullRefIfNotInitialized () [0x00000] in <6518d0e60ca84f6b915c07341a031a32>:0
  at System.Collections.Immutable.ImmutableArray`1[T].GetEnumerator () [0x00007] in <6518d0e60ca84f6b915c07341a031a32>:0
  at MonoDevelop.PackageManagement.PackageManagementCanReferenceProjectExtension.CanReferenceProject

The problem was that the project's TargetFramework setter is not
called when adding a new project to an existing solution so the
project's TargetFrameworkMonikers array was not initialized. This is
now handled.

Fixes VSTS #973344 - Unable to open Edit References dialog for new
project

mrward commented:
@monojenkins backport release-8.3

Therzok commented on main/src/core/MonoDevelop.Core/MonoDevelop.Projects/DotNetProject.cs:
```diff
@@ -444,7 +444,13 @@ public string StripImplicitNamespace (string ns)
+				// Adding a new project to an existing solution will not call the TargetFramework's setter which
+				// initializes this array, so handle this here if the array is not initialized.
+				if (targetFrameworkMonikers.IsDefaultOrEmpty)
+					targetFrameworkMonikers = ImmutableArray.Create<TargetFrameworkMoniker> (TargetFramework.Id);
```
Shouldn't we be doing this in `OnInitialize`?

Therzok replied:
See https://github.com/mono/monodevelop/blob/dffcf18d0387f7930640d7fbd34314cd4952c5be/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/DotNetProject.cs#L95

mrward replied:
This does not work. OnInitialize is called too early before the extension chain is initialized so trying to use DotNetProject.TargetFramework at this point throws an exception.

```
  System.InvalidOperationException : The extension chain can't be used before OnExtensionChainInitialized() method is called
  at MonoDevelop.Projects.WorkspaceObject.AssertExtensionChainCreated () [0x00043] in projects/monodevelop/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/WorkspaceObject.cs:379 
  at MonoDevelop.Projects.DotNetProject.get_ProjectExtension () [0x0000e] in projects/monodevelop/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/DotNetProject.cs:253 
  at MonoDevelop.Projects.DotNetProject.GetDefaultTargetFrameworkId () [0x00001] in projects/monodevelop/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/DotNetProject.cs:464 
  at MonoDevelop.Projects.DotNetProject.get_TargetFramework () [0x0000f] in projects/monodevelop/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/DotNetProject.cs:426 
  at MonoDevelop.Projects.DotNetProject.OnInitialize () [0x00061] in projects/monodevelop/main/src/core/MonoDevelop.Core/MonoDevelop.Projects/DotNetProject.cs:97 
```

Will see if there is any other method that can be used instead.

mrward replied:
OnItemReady might be something we can use instead. This is called when creating a new project and when loading an existing one. OnEndLoad is not called when creating a new project.

We may need to revisit what OnEndLoad is doing since this is initializing the file watcher - seems like that would not happen when adding a new C# project to an existing solution.

mrward replied:
Failed to get them both in the same method. Multi-target framework information needs to be available before the extension chain gets involved otherwise various tests break. But I cannot do the same for single target framework projects - so that bit is done in OnItemReady.

mrward committed:
[Core] Ensure target framework monikers updated when framework changed

The DotNetProject.TargetFrameworkMonikers were not updated if the
target framework was changed. Note that multi-target projects do not
allow the project's TargetFramework to be changed via the project
options dialog for a multi-target project so only update the monikers
if the project is a single target framework project.

Therzok reviewed (approved):
Looks great :+1:
